## AIPI 590

## Homework 2

## Jasmine Young

In this notebook I will compare the performance of my own implementation of Double DQN with the performance of stable-baseline3's traditional DQN.

In [1]:
#Install the modified code from my personal github
!pip install git+https://github.com/jgy4/stable-baselines3

  Cloning https://github.com/jgy4/stable-baselines3 to /tmp/pip-req-build-rf95lk69
  Running command git clone -q https://github.com/jgy4/stable-baselines3 /tmp/pip-req-build-rf95lk69


In [2]:
#Import Modules
import gym
import numpy as np
import matplotlib.pyplot as plt
from stable_baselines3 import DQN
from stable_baselines3.common.evaluation import evaluate_policy

In [3]:
#Create Two Mountain Car Environments
env = gym.make("MountainCar-v0")
env2 = gym.make("MountainCar-v0")

## Double DQN Model

In [4]:
#Intialize DoubleDQN Model
double_dqn_model = DQN("MlpPolicy",
            env,
            verbose=1,
            train_freq=16,
            gradient_steps=8,
            gamma=0.99,
            exploration_fraction=0.2,
            exploration_final_eps=0.07,
            target_update_interval=600,
            learning_starts=1000,
            buffer_size=10000,
            batch_size=128,
            learning_rate=4e-3,
            policy_kwargs=dict(net_arch=[256, 256]),
            seed=2, 
            doubleDQN=True)

Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.


In [5]:
#Train Double DQN Model
double_dqn_model.learn(int(2e5), log_interval=10)

----------------------------------
| rollout/            |          |
|    ep_len_mean      | 200      |
|    ep_rew_mean      | -200     |
|    exploration_rate | 0.954    |
| time/               |          |
|    episodes         | 10       |
|    fps              | 705      |
|    time_elapsed     | 2        |
|    total_timesteps  | 2000     |
| train/              |          |
|    learning_rate    | 0.004    |
|    loss             | 2.27e-05 |
|    n_updates        | 496      |
----------------------------------
----------------------------------
| rollout/            |          |
|    ep_len_mean      | 200      |
|    ep_rew_mean      | -200     |
|    exploration_rate | 0.907    |
| time/               |          |
|    episodes         | 20       |
|    fps              | 421      |
|    time_elapsed     | 9        |
|    total_timesteps  | 4000     |
| train/              |          |
|    learning_rate    | 0.004    |
|    loss             | 2.58e-06 |
|    n_updates      

In [6]:
#Evaluate The Training
mean_reward, std_reward = evaluate_policy(double_dqn_model, double_dqn_model.get_env(), deterministic=True, n_eval_episodes=20)
print(f"mean_reward:{mean_reward:.2f} +/- {std_reward:.2f}")

mean_reward:-127.05 +/- 15.27


## DQN Model

In [7]:
#Intialize DQN Model 
dqn_model = DQN("MlpPolicy",
            env2,
            verbose=1,
            train_freq=16,
            gradient_steps=8,
            gamma=0.99,
            exploration_fraction=0.2,
            exploration_final_eps=0.07,
            target_update_interval=600,
            learning_starts=1000,
            buffer_size=10000,
            batch_size=128,
            learning_rate=4e-3,
            policy_kwargs=dict(net_arch=[256, 256]),
            seed=2, 
            doubleDQN=False)

Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.


In [8]:
#Train DQN Model
dqn_model.learn(int(2e5), log_interval=10)

----------------------------------
| rollout/            |          |
|    ep_len_mean      | 200      |
|    ep_rew_mean      | -200     |
|    exploration_rate | 0.954    |
| time/               |          |
|    episodes         | 10       |
|    fps              | 937      |
|    time_elapsed     | 2        |
|    total_timesteps  | 2000     |
| train/              |          |
|    learning_rate    | 0.004    |
|    loss             | 0.00846  |
|    n_updates        | 496      |
----------------------------------
----------------------------------
| rollout/            |          |
|    ep_len_mean      | 200      |
|    ep_rew_mean      | -200     |
|    exploration_rate | 0.907    |
| time/               |          |
|    episodes         | 20       |
|    fps              | 626      |
|    time_elapsed     | 6        |
|    total_timesteps  | 4000     |
| train/              |          |
|    learning_rate    | 0.004    |
|    loss             | 0.000767 |
|    n_updates      

In [9]:
#Evaluate The Training
mean_reward, std_reward = evaluate_policy(dqn_model, dqn_model.get_env(), deterministic=True, n_eval_episodes=20)

print(f"mean_reward:{mean_reward:.2f} +/- {std_reward:.2f}")

mean_reward:-156.35 +/- 12.48


## Comparing DoubleDQN and DQN

The traditional DQN model performed with a mean reward of -156.35. My modified Double DQN Model performed with a mean reward of -127.05. Double DQN improved the performance of the model in the mountain car environment.